vcpkg.json

```diff
@@ -0,0 +1,17 @@
+{
+    "$schema": "https://raw.githubusercontent.com/microsoft/vcpkg/master/scripts/vcpkg.schema.json",
+    "name": "dosbox-staging",
+    "description": "DOSBox Staging is a fork of the DOSBox project that focuses on ease of use, modern technology and best practices.",
+    "version": "0.79.0",
+    "builtin-baseline": "19fafcdb3557f8a029ee1a4eb203600ff638448e",
+    "dependencies": [
+        {"name": "libpng", "version>=": "1.6.37#17"}, 
+        {"name": "sdl2", "version>=": "2.0.22#1"},
+        {"name": "sdl2-net", "version>=": "2.0.1#9"},
+        {"name": "libmt32emu", "version>=": "2.6.2"},
+        {"name": "opusfile", "version>=": "0.12#1"},
+        {"name": "fluidsynth", "version>=": "2.2.6#2"},
+        {"name": "gtest", "version>=": "1.11.0#5"},
+        {"name": "speexdsp", "version>=": "1.2.0#7"}
+    ]
```

Comment:
I'm not entirely sure on the best way to handle versions. The versioning docs are [here](https://github.com/microsoft/vcpkg/blob/master/docs/examples/versioning.getting-started.md).

From what I can gather, `builtin-baseline` is the absolute floor. Which makes me wonder if we set individual dependency versions, or rely on `builtin-baseline` to choose appropriate versions?

Reply:
Thanks for flagging how the manifest versioning works, @shermp.

We want CI giving us automatic exposure to the _latest-stable_ dependencies so we're aware of upgrades and can react and accommodate to changes, if needed.  The less manual intervention needed to "crank the version wheel", the better. (invariably, this becomes a time sink and a chore and those hours could have been spent on better things).

At the same time, being able to ratchet forward by setting an _absolute floor_ protects us from using old known-buggy dependencies; which is a similar mechanism we have in our Fedora manifest.

Looking at the documentation though, these two para's make me beleive we'll only ever the version as-specified, and not newer versions:

> Vcpkg is allowed to upgrade this constraint if a transitive dependency requires a newer version. For example, if zlib were to declare a dependency on fmt version 7.1.4 then vcpkg would install 7.1.4 instead.

> Vcpkg uses a minimum version approach, in our example, even if fmt version 8.0.0 were to be released, vcpkg would still install version 7.1.3#1 as that is the minimum version that satisfies the constraint. 

The only way we'll get newer version is through manual intervention:

> If you want to upgrade your dependencies, you can bump the minimum version constraint or use a newer builtin-baseline `{ "baseline": "3426db05b996481ca31e95fff3734cf23e0f51bc" }`

That's a lot more babysitting than I'm willing to put up with (and is exactly why so much software is riddled with vulnerabilities due to dependency version rot).

However, it then says:

> This field declares the versioning baseline for all ports. Setting a baseline is required to enable versioning, otherwise you will get the current versions on the ports directory. 

So maybe we can delete the `builtin-baseline` and get this "otherwise you will get the current versions on the ports directory." scenario?

Reply:
If that's the way you want to do it, sure.

My only concern is that a dependency is upgraded to a new version which breaks API/ABI.

Reply:
> My only concern is that a dependency is upgraded to a new version which breaks API/ABI.

Yes; I prefer that.

I'd much rather have automatic forces at-play driving the code-base to stay current, against the latest-stable dependencies and APIs, with zero effort on our part, and be notified at our earliest convenience. 

Contrast this with the rote elbow grease method of checking for new versions and updating manifest records:
 - First of all, someone needs to care enough to periodically check this stuff.
 - Let's say some months have passed and someone decides to roll the manifests forward.
 - Let's say a bunch of libs have updated in some rather painful ways, so CI is now hit with various API breakages. Perhaps also, some libs are having architecture-specific problems as well. 
 - The person who rolled the manifest forward now has two paths:
     - Lump it an embark on fixing all this stuff (waterfall + big-bang approach)
     - Run back to the previously working manifest (ignore it, largest big-bang down the road)

I've seen a lot of companies choose the 2nd path. 

But if they do decide to fix stuff, then the end result for both approaches is the same amount of work: however, the first approach "early and automatic notification", provides change in the smallest possible increment on a per-library basis. 

I'd much rather automatic and early notification of API breaks, through no effort of our own, than rely on manual checks and risking an even bigger-bang fix effort.

Worst case: even in the automatic path, one could then decide to freeze a library back to the prior version.

Reply:
Done.

Reply:
Thanks, @shermp.

For tagged releases, locking in the current `builtin-baseline` makes a lot of sense, so we can go back in time and build those binaries in a reproduceable way.  I haven't checked, but if you can leave the `builtin-baseline` commented out, then we can add that commit-hash just for releases.

I see benefits in both paths: if we can have both, then that's ideal :+1:

Reply:
`vcpkg.json` is a strict JSON file, no comments allowed.

Reply:
Maybe add what you've got above to a `vcpkg.json.populate_for_releases_only` helper file? (or something like that?)

Reply:
Then we can add a step in the Wiki release steps to populate and rename that fil with the current vcpkg checksum and latest versions.

Reply:
Where would you want such a file? And should it contain the package versions? Or just the baseline?

Reply:
In it's active location, but just renamed to become inactive. 
(Similar approach to Linux records in /etc/defaults/ and other locations.. the example files are all "in place ready to go", but renamed to be inactive)

Reply:
It's a bit tedious, although could probably be scripted.

I'll add the whole thing for now, when it comes release time, we can decide what to do.

Reply:
Perfect; thanks, @shermp !
The fewer details we need to manage, the better. If that means we only need to drop in the baseline, then that's perfect.

  If someone needs to get forensic - they can use the commit hash to reproduce the build,  drill through logs, and investigate commit histories in vcpkg. So probably it's good enough.

Reply:
Ok, done